File: ros2_ws/src/omnibot_control/omnibot_control/encoder_node.cpp
#include <lgpio.h>

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/int32_multi_array.hpp>

#include <chrono>
// #include <memory>
#include <vector>

class EncoderReader {
public:
    EncoderReader(int handle, int pin_a, int pin_b)
        : pin_a_(pin_a), pin_b_(pin_b), counter_(0), last_a_(0), last_b_(0) {
        lgGpioClaimInput(handle, LG_SET_PULL_UP, pin_a_);
        lgGpioClaimInput(handle, LG_SET_PULL_UP, pin_b_);

        last_a_ = gpioRead(pin_a_);
        last_b_ = gpioRead(pin_b_);

        callback_a_ = gpioSetAlertFuncEx(pin_a_, &EncoderReader::_pulseCallback, this);
        callback_b_ = gpioSetAlertFuncEx(pin_b_, &EncoderReader::_pulseCallback, this);
    }

    ~EncoderReader() {
        gpioSetAlertFuncEx(pin_a_, nullptr, nullptr);
        gpioSetAlertFuncEx(pin_b_, nullptr, nullptr);
    }

    long read() const {
        return counter_;
    }

    void reset() {
        counter_ = 0;
    }

private:
    static void _pulseCallback(int gpio, int level, uint32_t tick, void* userdata) {
        EncoderReader* self = static_cast<EncoderReader*>(userdata);
        self->_onPulse(gpio, level);
    }

    void _onPulse(int gpio, int level) {
        if (gpio == pin_a_)
            last_a_ = level;
        else if (gpio == pin_b_)
            last_b_ = level;

        if (gpio == pin_a_ && level == 1) {
            if (last_b_ == 1)
                counter_++;
            else
                counter_--;
        } else if (gpio == pin_b_ && level == 1) {
            if (last_a_ == 1)
                counter_--;
            else
                counter_++;
        }
    }

    int pin_a_, pin_b_;
    volatile long counter_;
    volatile int last_a_, last_b_;
    int callback_a_, callback_b_;
};

class EncoderNode : public rclcpp::Node {
public:
    EncoderNode() : Node("encoder") {
        this->declare_parameter("encoder1.pin_a", 10);
        this->declare_parameter("encoder1.pin_b", 9);
        this->declare_parameter("encoder2.pin_a", 13);
        this->declare_parameter("encoder2.pin_b", 19);
        this->declare_parameter("encoder3.pin_a", 20);
        this->declare_parameter("encoder3.pin_b", 21);

        int e1a = this->get_parameter("encoder1.pin_a").as_int();
        int e1b = this->get_parameter("encoder1.pin_b").as_int();
        int e2a = this->get_parameter("encoder2.pin_a").as_int();
        int e2b = this->get_parameter("encoder2.pin_b").as_int();
        int e3a = this->get_parameter("encoder3.pin_a").as_int();
        int e3b = this->get_parameter("encoder3.pin_b").as_int();

        if (lgGpiochipOpen(4) < 0) {
            RCLCPP_ERROR(this->get_logger(), "Error initilize GPIO");
            rclcpp::shutdown();
        }



        encoders_.push_back(std::make_unique<EncoderReader>(e1a, e1b));
        encoders_.push_back(std::make_unique<EncoderReader>(e2a, e2b));
        encoders_.push_back(std::make_unique<EncoderReader>(e3a, e3b));

        publisher_ = this->create_publisher<std_msgs::msg::Int32MultiArray>("/encoder_ticks", 10);

        timer_ = this->create_wall_timer(std::chrono::milliseconds(20),
                                         std::bind(&EncoderNode::timerCallback, this));

        RCLCPP_INFO(this->get_logger(), "Encoder Node запущена. Публикуем в /encoder_ticks");
    }

    ~EncoderNode() {
        lgGpiochipClose(4);
    }

private:
    void timerCallback() {
        auto msg = std_msgs::msg::Int32MultiArray();
        msg.data.resize(3);
        for (size_t i = 0; i < encoders_.size(); ++i) {
            msg.data[i] = encoders_[i]->read();
        }
        publisher_->publish(msg);
    }

    std::vector<std::unique_ptr<EncoderReader>> encoders_;
    rclcpp::Publisher<std_msgs::msg::Int32MultiArray>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<EncoderNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/omnibot_control/omnibot_control/encoder_node.py
import lgpio
from typing import Any
import threading

import rclpy
from rclpy.node import Node
from std_msgs.msg import Int32MultiArray

class RotaryEncoder:
    def __init__(self, chip: Any, pin_a: int, pin_b: int):
        self.handle = chip
        self.pin_a = pin_a
        self.pin_b = pin_b
        self.counter: int = 0
        self.lock = threading.Lock()

        lgpio.gpio_claim_input(self.handle, self.pin_a)
        lgpio.gpio_claim_input(self.handle, self.pin_b)

        self.last_a = lgpio.gpio_read(self.handle, self.pin_a)
        self.last_b = lgpio.gpio_read(self.handle, self.pin_b)

        self.cb_a = lgpio.callback(self.handle, self.pin_a, lgpio.BOTH_EDGES, self._on_change)
        self.cb_b = lgpio.callback(self.handle, self.pin_b, lgpio.BOTH_EDGES, self._on_change)

    def _on_change(self, chip, gpio, level, timestamp):
        a = lgpio.gpio_read(self.handle, self.pin_a)
        b = lgpio.gpio_read(self.handle, self.pin_b)

        if a == self.last_a and b == self.last_b:
            return
        
        delta = 0
        if self.last_a == 0 and self.last_b == 0:
            if a == 0 and b == 1:
                delta = -1
            elif a == 1 and b == 0:
                delta = +1
        elif self.last_a == 0 and self.last_b == 1:
            if a == 1 and b == 1:
                delta = -1
            elif a == 0 and b == 0:
                delta = +1
        elif self.last_a == 1 and self.last_b == 0:
            if a == 0 and b == 0:
                delta = -1
            elif a == 1 and b == 1:
                delta = +1
        elif self.last_a == 1 and self.last_b == 1:
            if a == 1 and b == 0:
                delta = -1
            elif a == 0 and b == 1:
                delta = +1
                
        if delta != 0:
            with self.lock:
                self.counter += delta

        self.last_a = a
        self.last_b = b

    def count(self) -> int:
        with self.lock:
            return self.counter

    def reset(self):
        with self.lock:
            self.counter = 0

    def stop(self):
        self.cb_a.cancel()
        self.cb_b.cancel()

class EncoderNode(Node):
    def __init__(self):
        super().__init__('encoder_node')

        self._log_counter = 0

        self.chip = lgpio.gpiochip_open(4)

        self.encoders = [
            RotaryEncoder(self.chip, 10, 9),
            RotaryEncoder(self.chip, 13, 19),
            RotaryEncoder(self.chip, 20, 21)
        ]

        self.pub = self.create_publisher(
            Int32MultiArray,
            '/encoders_counter',
            10
        )

        self.timer = self.create_timer(0.05, self.publish_count)

    def publish_count(self):
        msg = Int32MultiArray()
        counts = []
        for enc in self.encoders:
            counts.append(enc.count())
            enc.reset()
        # counts = [enc.count() for enc in self.encoders]
        msg.data = counts
        self.pub.publish(msg)

        self._log_counter += 1
        if self._log_counter % 40 == 0:
            self.get_logger().info(f"Состояние энкодеров: {counts}")

    def destroy_node(self) -> bool:
        for enc in self.encoders:
            enc.stop()
        lgpio.gpiochip_close(self.chip)
        return super().destroy_node()

def main(args=None) -> None:
    rclpy.init(args=args)
    node = EncoderNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()